lib/blacklight_advanced_search.rb

```diff
@@ -12,37 +12,37 @@ module BlacklightAdvancedSearch
   require 'blacklight_advanced_search/engine'
 
   # Utility method used in our solr search logic.
-  # Merges new_hash into source_hash, but will recursively
-  # merge nested arrays and hashes too; also will NOT merge nil
-  # or blank values from new_hash into source_hash, nil or blank values
-  # in new_hash will not overwrite values in source_hash.
-  def self.deep_merge!(source_hash, new_hash)
-    # We used to use built-in source_hash.merge() with a block arg
-    # to customize merge behavior, but that was breaking in some
-    # versions of BL/Rails where source_hash was a kind of HashWithIndifferentAccess,
-    # and hwia is unreliable in some versions of Rails. Oh well.
-    # https://github.com/projectblacklight/blacklight/issues/827
+  # Like Rails Hash#deep_merge, merges 2 hashes recursively, including nested Arrays and Hashes.
+  # Unlike Rails Hash#deep_merge:
+  # - will NOT merge nil values over existing ones
+  # - will NOT merge (non-FalseClass) blank values
+  # - WILL deduplicate values from arrays after merging them
+  #
+  # @param [Hash|HashWithIndifferentAccess] source_hash
+  # @param [Hash|HashWithIndifferentAccess] new_hash
+  # @return [Hash] the deeply merged hash
+  # @see Rails #deep_merge http://apidock.com/rails/v4.2.1/Hash/deep_merge
+  # @example new_hash = BlacklightAdvancedSearch.deep_merge(h1, h2)
+  def self.deep_merge(source_hash, new_hash)
+    source_hash.deep_merge(new_hash) { |x, y, z| merge_conflict_resolution(x, y, z) }
+  end
 
-    new_hash.each_pair do |key, new_value|
-      old = source_hash.fetch(key, nil)
+  # this one side-effects the first param
+  # @see #deep_merge
+  # @deprecated use `new_hash = BlacklightAdvancedSearch.deep_merge(h1, h2)` instead
+  def self.deep_merge!(source_hash, new_hash)
+    source_hash.deep_merge!(new_hash) { |x, y, z| merge_conflict_resolution(x, y, z) }
```

Comment:
Can we issue a proper deprecation warning? I'm fine with either pulling in `deprecation` or being lazy and just using `ActiveSupport::Deprecation.warn` directly.